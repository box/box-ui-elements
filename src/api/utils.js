/**
 * @flow
 * @file Utils for the box APIs
 * @author Box
 */

import type { Comment } from '../common/types/feed';

/**
 * Formats comment data (including replies) for use in components.
 *
 * @param {Comment} comment - An individual comment entry from the API
 * @return {Comment} Updated comment
 */
export const formatComment = (comment: Comment): Comment => {
    const formattedComment = {
        ...comment,
        tagged_message: comment.message,
    };

    if (comment.replies && comment.replies.length) {
        formattedComment.replies = comment.replies.map(formatComment);
    }

    return formattedComment;
};

export default {
    formatComment,
};
