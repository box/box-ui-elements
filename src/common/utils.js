// @flow
import isFinite from 'lodash/isFinite';

const isFloat = (value: number) => {
    return !/^\s*$/.test(String(value)) && isFinite(value);
};

// eslint-disable-next-line import/prefer-default-export
export { isFloat };
