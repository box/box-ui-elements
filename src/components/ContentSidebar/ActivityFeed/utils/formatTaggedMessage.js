/**
 * @flow
 * @file Util for formatting tagged messages
 */

import * as React from 'react';
import { Link } from 'box-react-ui/lib/components/link';
import Mention from '../comment/Mention';

// this regex matches one of the following regular expressions:
// mentions: ([@＠﹫]\[[0-9]+:[^\]]+])
// urls: (?:\b)((?:(?:ht|f)tps?:\/\/)[\w\._\-]+(:\d+)?(\/[\w\-_\.~\+\/#\?&%=:\[\]@!$'\(\)\*;,]*)?)
// NOTE: There are useless escapes in the regex below, should probably remove them when safe
// eslint-disable-next-line
const splitRegex = /((?:[@＠﹫]\[[0-9]+:[^\]]+])|(?:\b(?:(?:ht|f)tps?:\/\/)[\w\._\-]+(?::\d+)?(?:\/[\w\-_\.~\+\/#\?&%=:\[\]@!$'\(\)\*;,]*)?))/gim;

/**
 * Formats a message a string and replaces the following:
 * - all occurrence of mention patterns with a Mention component
 * - all occurrence of urls with a Link component
 * Ex mention format: @[123:Hello World]
 * @param {String} tagged_message The message string to format
 * @param {String} itemID The id of the tagged message
 * @param {Boolean} shouldReturnString The boolean value whether it should return string
 * @returns {String|React.Node}
 */
const formatTaggedMessage = (
    tagged_message: string,
    itemID: string,
    shouldReturnString: boolean
): React.Node | string => {
    const contentItems = tagged_message.split(splitRegex).map((text: string, contentIndex: number) => {
        const contentKey = `${contentIndex}-${itemID}`;
        // attempt mention match
        const mentionMatch = text.match(/([@＠﹫])\[([0-9]+):([^\]]+)]/i);
        if (mentionMatch) {
            const [, trigger, id, name] = mentionMatch;
            if (shouldReturnString) {
                return `${trigger}${name}`;
            }
            return <Mention id={Number(id)} key={contentKey}>{`${trigger}${name}`}</Mention>;
        }

        if (!shouldReturnString) {
            // attempt url match
            // NOTE: There are useless escapes in the regex below, should probably remove them when safe
            const urlMatch = text.match(
                // eslint-disable-next-line no-useless-escape
                /((?:(?:ht|f)tps?:\/\/)[\w\._\-]+(?::\d+)?(?:\/[\w\-_\.~\+\/#\?&%=:\[\]@!$'\(\)\*;,]*)?)/i
            );
            if (urlMatch) {
                const [, url] = urlMatch;
                return (
                    <Link key={contentKey} href={url}>
                        {url}
                    </Link>
                );
            }
        }

        return text;
    });

    return shouldReturnString ? contentItems.join('') : contentItems;
};

export default formatTaggedMessage;
