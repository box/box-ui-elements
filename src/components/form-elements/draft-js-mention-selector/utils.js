// @flow
import { EditorState, Modifier } from 'draft-js';

export type Mention = {
    blockID: string,
    end: number,
    mentionString: string,
    mentionTrigger: string,
    start: number,
};

const defaultMentionTriggers = ['@', '＠', '﹫'];
const defaultMentionTriggersString = defaultMentionTriggers.reduce((prev, current) => `${prev}\\${current}`, '');
const defaultMentionPattern = new RegExp(`([${defaultMentionTriggersString}])([^${defaultMentionTriggersString}]*)$`);

/**
 * Extracts the active mention from the editor state
 */
function getActiveMentionForEditorState(
    editorState: EditorState,
    mentionPattern: RegExp = defaultMentionPattern,
): Mention | null {
    const contentState = editorState.getCurrentContent();
    const selectionState = editorState.getSelection();

    const startKey = selectionState.getStartKey();
    const activeBlock = contentState.getBlockForKey(startKey);

    const cursorPosition = selectionState.getStartOffset();

    let result = null;

    // Break the active block into entity ranges.
    activeBlock.findEntityRanges(
        character => character.getEntity() === null,
        (start, end) => {
            // Find the active range (is the cursor inside this range?)
            if (start <= cursorPosition && cursorPosition <= end) {
                // Determine if the active range contains a mention.
                const activeRangeText = activeBlock.getText().substr(start, cursorPosition - start);
                const mentionMatch = activeRangeText.match(mentionPattern);

                if (mentionMatch) {
                    result = {
                        blockID: startKey,
                        mentionString: mentionMatch[2],
                        mentionTrigger: mentionMatch[1],
                        start: start + mentionMatch.index,
                        end: cursorPosition,
                    };
                }
            }

            return null;
        },
    );

    return result;
}

/**
 * Inserts a selected mention into the editor
 */
function addMention(editorState: EditorState, activeMention: Mention | null, mention: Object): EditorState {
    const { start, end } = activeMention || {};

    const { id, name } = mention;

    const contentState = editorState.getCurrentContent();
    const selectionState = editorState.getSelection();

    const preInsertionSelectionState = selectionState.merge({
        anchorOffset: start,
        focusOffset: end,
    });

    const textToInsert = `@${name}`;

    const contentStateWithEntity = contentState.createEntity('MENTION', 'IMMUTABLE', { id });

    const entityKey = contentStateWithEntity.getLastCreatedEntityKey();

    const contentStateWithLink = Modifier.replaceText(
        contentState,
        preInsertionSelectionState,
        textToInsert,
        null,
        entityKey,
    );

    const spaceOffset = preInsertionSelectionState.getStartOffset() + textToInsert.length;
    const selectionStateForAddingSpace = preInsertionSelectionState.merge({
        anchorOffset: spaceOffset,
        focusOffset: spaceOffset,
    });

    const contentStateWithLinkAndExtraSpace = Modifier.insertText(
        contentStateWithLink,
        selectionStateForAddingSpace,
        ' ',
    );

    const editorStateWithLink = EditorState.push(editorState, contentStateWithLinkAndExtraSpace, 'change-block-type');

    return editorStateWithLink;
}

/**
 * Formats the editor's text such that it will be accepted by the server.
 */
function getFormattedCommentText(editorState: EditorState): { hasMention: boolean, text: string } {
    const contentState = editorState.getCurrentContent();
    const blockMap = contentState.getBlockMap();

    const resultStringArr = [];

    // The API needs to explicitly know if a message contains a mention.
    let hasMention = false;

    // For all ContentBlocks in the ContentState:
    blockMap.forEach(block => {
        const text = block.getText();
        const blockMapStringArr = [];

        // Break down the ContentBlock into ranges
        block.findEntityRanges(
            () => true,
            (start, end) => {
                const entityKey = block.getEntityAt(start);
                // If the range is an Entity, format its text eg "@[1:Username]"
                // Otherwise append its text to the block result as-is
                if (entityKey) {
                    const entity = contentState.getEntity(entityKey);
                    const stringToAdd = `@[${entity.getData().id}:${text.substring(start + 1, end)}]`;
                    blockMapStringArr.push(stringToAdd);
                    hasMention = true;
                } else {
                    blockMapStringArr.push(text.substring(start, end));
                }
            },
        );
        resultStringArr.push(blockMapStringArr.join(''));
    });

    // Concatenate the array of block strings with newlines
    // (Each block represents a paragraph)
    return { text: resultStringArr.join('\n'), hasMention };
}

export {
    addMention,
    defaultMentionTriggers,
    defaultMentionPattern,
    getActiveMentionForEditorState,
    getFormattedCommentText,
};
