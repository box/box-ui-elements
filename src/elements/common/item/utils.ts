import { FILE_EXTENSIONS } from './constants';

const getFileIconType = (extension = '') => {
    const iconTypes = {};

    Object.entries(FILE_EXTENSIONS).forEach(([iconType, typeExtensions]) => {
        typeExtensions.forEach(typeExtension => {
            iconTypes[typeExtension] = iconType;
        });
    });

    return iconTypes[extension] || 'default';
};

export { getFileIconType };
