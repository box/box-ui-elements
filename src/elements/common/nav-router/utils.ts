import { History } from 'history';
import { WithNavRouterProps } from './types';

export interface MatchResult {
    isExact: boolean;
    params: Record<string, string>;
}

export function matchPath(pathname: string, path: string | string[]): MatchResult {
    const paths = Array.isArray(path) ? path : [path];
    const params: Record<string, string> = {};

    for (const currentPath of paths) {
        const pathSegments = currentPath.split('/').filter(Boolean);
        const pathnameSegments = pathname.split('/').filter(Boolean);

        if (pathSegments.length <= pathnameSegments.length) {
            let isMatch = true;
            const segmentLength = pathSegments.length;

            for (let i = 0; i < segmentLength; i += 1) {
                const pathSegment = pathSegments[i];
                const pathnameSegment = pathnameSegments[i];

                if (pathSegment.startsWith(':')) {
                    // This is a parameter
                    const paramName = pathSegment.slice(1);
                    params[paramName] = pathnameSegment;
                } else if (pathSegment === '*') {
                    // Wildcard matches anything
                    params[i.toString()] = pathnameSegments.slice(i).join('/');
                    break;
                } else if (pathSegment !== pathnameSegment) {
                    isMatch = false;
                    break;
                }
            }

            if (isMatch) {
                return {
                    isExact: pathSegments.length === pathnameSegments.length,
                    params,
                };
            }
        }
    }

    return { isExact: false, params: {} };
}

export function useMatch(
    props: Partial<WithNavRouterProps> & { history: History },
    path: string | string[],
): MatchResult {
    const location = props.location || props.history.location;
    return matchPath(location.pathname, path);
}
