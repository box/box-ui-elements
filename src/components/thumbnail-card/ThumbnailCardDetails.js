// @flow
import * as React from 'react';

import Tooltip from '../tooltip';
import { useIsContentOverflowed } from '../../utils/dom';

type Props = {
    actionItem?: React.Element<any>,
    icon?: React.Node,
    subtitle?: React.Node,
    title: React.Node,
};

type TitleProps = {
    title: React.Node,
};

const Title = ({ title }: TitleProps) => {
    // $FlowFixMe
    const textRef: { current: null | HTMLElement } = React.useRef(null);
    const isTextOverflowed = useIsContentOverflowed(textRef);

    return (
        <Tooltip isDisabled={!isTextOverflowed} text={title}>
            <div ref={textRef} className="thumbnail-card-title">
                {title}
            </div>
        </Tooltip>
    );
};

const ThumbnailCardDetails = ({ actionItem, icon, subtitle, title }: Props) => (
    <div className="thumbnail-card-details">
        {icon}
        <div className="thumbnail-card-details-content">
            <div className="ThumbnailCardDetails-bodyText">
                <Title title={title} />
                {subtitle && <div className="thumbnail-card-subtitle">{subtitle}</div>}
            </div>
            {actionItem}
        </div>
    </div>
);

export default ThumbnailCardDetails;
