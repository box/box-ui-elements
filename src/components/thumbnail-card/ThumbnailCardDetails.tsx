import * as React from 'react';

import Tooltip from '../tooltip';
import { useIsContentOverflowed } from '../../utils/dom';

export interface ThumbnailCardDetailsProps {
    actionItem?: React.ReactElement;
    icon?: React.ReactNode;
    onKeyDown?: () => void;
    subtitle?: React.ReactNode;
    title: React.ReactNode;
}

interface TitleProps {
    onKeyDown?: () => void;
    title: React.ReactNode;
}

const Title = ({ title, onKeyDown }: TitleProps) => {
    const textRef = React.useRef<HTMLDivElement>(null);
    const isTextOverflowed = useIsContentOverflowed(textRef);

    return (
        <Tooltip isDisabled={!isTextOverflowed} text={title}>
            {/* eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex, jsx-a11y/no-static-element-interactions */}
            <div ref={textRef} role="link" className="thumbnail-card-title" tabIndex={0} onKeyDown={onKeyDown}>
                {title}
            </div>
        </Tooltip>
    );
};

const ThumbnailCardDetails = ({ actionItem, icon, subtitle, title, onKeyDown }: ThumbnailCardDetailsProps) => (
    <div className="thumbnail-card-details">
        {icon}
        <div className="thumbnail-card-details-content">
            <div className="ThumbnailCardDetails-bodyText">
                <Title title={title} onKeyDown={onKeyDown} />
                {subtitle && <div className="thumbnail-card-subtitle">{subtitle}</div>}
            </div>
            {actionItem}
        </div>
    </div>
);

export default ThumbnailCardDetails;
