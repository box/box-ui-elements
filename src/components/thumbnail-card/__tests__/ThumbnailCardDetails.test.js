// @flow
import * as React from 'react';

import { mount, shallow } from 'enzyme';

import * as libDom from '../../../utils/dom';

import ThumbnailCardDetails from '../ThumbnailCardDetails';

const getWrapper = (props = {}) => shallow(<ThumbnailCardDetails title={<div>Foo Bar!</div>} {...props} />);

jest.mock('../../../utils/dom', () => ({ useIsContentOverflowed: jest.fn() }));

describe('components/thumbnail-card/ThumbnailCardDetails', () => {
    beforeEach(() => {
        libDom.useIsContentOverflowed.mockReturnValue(false);
    });

    test('should render', () => {
        const wrapper = getWrapper();

        expect(wrapper).toMatchSnapshot();
    });

    test('should render icon', () => {
        const icon = <img alt="icon" />;
        const wrapper = getWrapper({ icon });

        expect(wrapper).toMatchSnapshot();
    });

    test('should render subtitle', () => {
        const subtitle = <div>Subtitle!</div>;
        const wrapper = getWrapper({ subtitle });

        expect(wrapper).toMatchSnapshot();
    });

    test('should render actionItem', () => {
        const actionItem = <button type="button">Click Me</button>;
        const wrapper = getWrapper({ actionItem });

        expect(wrapper).toMatchSnapshot();
    });

    test('should render a Tooltip if text is overflowed', () => {
        libDom.useIsContentOverflowed.mockReturnValue(true);

        const wrapper = mount(<ThumbnailCardDetails title={<div>Foo Bar!</div>} />);

        expect(wrapper.find('Tooltip').length).toBe(1);
    });
});
