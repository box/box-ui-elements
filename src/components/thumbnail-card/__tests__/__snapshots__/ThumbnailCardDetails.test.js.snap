// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/thumbnail-card/ThumbnailCardDetails should render 1`] = `
<div
  className="thumbnail-card-details"
>
  <div
    className="thumbnail-card-details-content"
  >
    <div
      className="ThumbnailCardDetails-bodyText"
    >
      <Title
        title={
          <div>
            Foo Bar!
          </div>
        }
      />
    </div>
  </div>
</div>
`;

exports[`components/thumbnail-card/ThumbnailCardDetails should render actionItem 1`] = `
<div
  className="thumbnail-card-details"
>
  <div
    className="thumbnail-card-details-content"
  >
    <div
      className="ThumbnailCardDetails-bodyText"
    >
      <Title
        title={
          <div>
            Foo Bar!
          </div>
        }
      />
    </div>
    <button
      type="button"
    >
      Click Me
    </button>
  </div>
</div>
`;

exports[`components/thumbnail-card/ThumbnailCardDetails should render icon 1`] = `
<div
  className="thumbnail-card-details"
>
  <img
    alt="icon"
  />
  <div
    className="thumbnail-card-details-content"
  >
    <div
      className="ThumbnailCardDetails-bodyText"
    >
      <Title
        title={
          <div>
            Foo Bar!
          </div>
        }
      />
    </div>
  </div>
</div>
`;

exports[`components/thumbnail-card/ThumbnailCardDetails should render subtitle 1`] = `
<div
  className="thumbnail-card-details"
>
  <div
    className="thumbnail-card-details-content"
  >
    <div
      className="ThumbnailCardDetails-bodyText"
    >
      <Title
        title={
          <div>
            Foo Bar!
          </div>
        }
      />
      <div
        className="thumbnail-card-subtitle"
      >
        <div>
          Subtitle!
        </div>
      </div>
    </div>
  </div>
</div>
`;
